File: codes/8psk/mats.py
#Code by GVV Sharma
#July 10, 2020
#Released under GNU/GPL

#8PSK constellation 
#Demodulation matrx
#Qfunction 

import numpy as np
from scipy import special

#Generating costellation points
s = np.zeros((8,2))
for i in range(8):
	s[i,:] = np.array(([np.cos(i*2*np.pi/8),np.sin(i*2*np.pi/8)]))

#Generating demodulation matrix
A = np.zeros((8,2,2))
A[0,:,:] = np.array(([np.sqrt(2)-1,1],[np.sqrt(2)-1,-1]))
A[1,:,:] = np.array(([np.sqrt(2)+1,-1],[-(np.sqrt(2)-1),1]))
A[2,:,:] = np.array(([-(np.sqrt(2)+1),1],[np.sqrt(2)+1,1]))
A[3,:,:] = np.array(([np.sqrt(2)-1,1],[-(np.sqrt(2)+1),-1]))
A[4,:,:] = np.array(([-(np.sqrt(2)-1),-1],[-(np.sqrt(2)-1),1]))
A[5,:,:] = np.array(([-(np.sqrt(2)+1),1],[np.sqrt(2)-1,-1]))
A[6,:,:] = np.array(([np.sqrt(2)+1,-1],[-(np.sqrt(2)+1),-1]))
A[7,:,:] = np.array(([-(np.sqrt(2)-1),-1],[np.sqrt(2)+1,1]))

#Gray code
gray = np.zeros((8,3))
gray[0,:] = np.array(([0,0,0]))
gray[1,:] = np.array(([0,0,1]))
gray[2,:] = np.array(([0,1,1]))
gray[3,:] = np.array(([0,1,0]))
gray[4,:] = np.array(([1,1,0]))
gray[5,:] = np.array(([1,1,1]))
gray[6,:] = np.array(([1,0,1]))
gray[7,:] = np.array(([1,0,0]))


#Q-function
def qfunc(x):
	return 0.5*special.erfc(x/np.sqrt(2))

File: codes/MMSE/mats.m
#Code by GVV Sharma
#August 1, 2020

#8PSK constellation 
#Demodulation matrx


#Generating constellation points
s = zeros(8,2);
s_comp = zeros(8,1);
for i =1:8
	s(i,:) = [cos(i*2*pi/8) sin(i*2*pi/8)]; #vector
	s_comp(i) = s(i,1)+1j*s(i,2); #equivalent complex number
end
#
#Generating demodulation matrix
A = zeros(2,2,8);
A(:,:,1) = [sqrt(2)-1 1; sqrt(2)-1 -1];
A(:,:,2) = [sqrt(2)+1 -1; -(sqrt(2)-1) 1];
A(:,:,3) = [-(sqrt(2)+1) 1; sqrt(2)+1 1];
A(:,:,4) = [sqrt(2)-1 1; -(sqrt(2)+1) -1];
A(:,:,5) = [-(sqrt(2)-1) -1; -(sqrt(2)-1) 1];
A(:,:,6) = [-(sqrt(2)+1) 1; sqrt(2)-1 -1];
A(:,:,7) = [sqrt(2)+1 -1; -(sqrt(2)+1) -1];
A(:,:,8) = [-(sqrt(2)-1) -1; sqrt(2)+1 1];

#Gray code
gray = zeros(8,3);
gray(1,:) = [0 0 0];
gray(2,:) = [0 0 1];
gray(3,:) = [0 1 1];
gray(4,:) = [0 1 0];
gray(5,:) = [1 1 0];
gray(6,:) = [1 1 1];
gray(7,:) = [1 0 1];
gray(8,:) = [1 0 0];
#
#gray
#
##Q-function
#def qfunc(x):
#	return 0.5*special.erfc(x/sqrt(2
